src/amg/azext_amg/restore.py

```diff
@@ -121,83 +144,211 @@ def _create_library_panel(grafana_url, file_path, http_headers):
-            }
-            (status, content) = send_grafana_patch(f'{grafana_url}/api/library-elements/{uid}',
-                                                   json.dumps(patch_payload), http_headers)
-    print_styled_text([
-        (Style.WARNING, f'Create library panel {panel_name}: '),
-        (Style.SUCCESS, 'SUCCESS') if status == 200 else (Style.ERROR, 'FAILURE')
-    ])
+        if overwrite:
+            uid = payload['uid']
+            panel_uri = f'{grafana_url}/api/library-elements/{uid}'
+            (status, content) = send_grafana_get(panel_uri, http_headers)
+            if status == 200:
+                patch_payload = {
+                    'name': panel_name,
+                    'model': payload['model'],
+                    'version': content['result']['version'],
+                    'kind': payload['kind']
+                }
+                (status, content) = send_grafana_patch(f'{grafana_url}/api/library-elements/{uid}',
+                                                       json.dumps(patch_payload), http_headers)
+
+            print_styled_text([
+                (Style.WARNING, f'Overwrite library panel {panel_name}: '),
+                (Style.SUCCESS, 'SUCCESS') if status == 200 else (Style.ERROR, 'FAILURE')
+            ])
+        else:
+            print_styled_text([
+                (Style.WARNING, f'Create library panel {panel_name}: '),
+                (Style.ERROR, 'FAILURE'),
+                (Style.ERROR, ' (name or UID already exists,'),
+                (Style.ERROR, ' please enable --overwrite if you want to overwrite the library panel)')
+            ])
+
+    else:
+        print_styled_text([
+            (Style.WARNING, f'Create library panel {panel_name}: '),
+            (Style.SUCCESS, 'SUCCESS') if status == 200 else (Style.ERROR, 'FAILURE')
+        ])
+
     logger.info("status: %s, msg: %s", status, content)
+    return status == 200
+
+
+def check_library_panel_exists(grafana_url, payload, http_headers):
+    (status, _) = send_grafana_get(f'{grafana_url}/api/library-elements/{payload["uid"]}', http_headers)
+    return status == 200
 
 
 # Restore snapshots
-def _create_snapshot(grafana_url, file_path, http_headers):
+def _load_and_create_snapshot(grafana_url, file_path, http_headers):
     with open(file_path, 'r', encoding="utf8") as f:
         data = f.read()
 
     snapshot = json.loads(data)
+    create_snapshot(grafana_url, snapshot, http_headers, overwrite=True)
+
+
+def check_snapshot_exists(grafana_url, snapshot_key, http_headers):
+    (status, _) = get_snapshot(snapshot_key, grafana_url, http_headers)
+    return status == 200
+
+
+def create_snapshot(grafana_url, snapshot, http_headers, overwrite):
     try:
         snapshot['name'] = snapshot['dashboard']['title']
     except KeyError:
         snapshot['name'] = "Untitled Snapshot"
+    snapshot_name = snapshot['name']
+
+    exists_before = False
+    # TODO: in backup we don't save the key.
```

Comment:
so we're missing this functionality at the moment?

Reply:
Yeah, I did not change the functionality in the backup functionality (we didn't store the key in the backup) but we needed to save key for migrate so we can use it to help check if the snapshot exists before (since snapshots don't have uids, keys are kinda used for that here).